### ISSUE TITLE
Send streams

### ISSUE THREAD

**Issue by joonnna** [gwierzchowski/csv-async#2]
I'm currently trying to return `DeserializeRecordsIntoStream` as a boxed trait object that will be used in a multi-threaded runtime.
Something like: 
```
Box<dyn Stream<Item = Result<T, StreamError>> + Unpin + Send>
```
However, the `DeserializeRecordsIntoStream` is `!Send`, I can't seem to find a workaround on how to use the stream in an environment where `Send` is required.
Do you have any suggested workaround?

I'm not familiar with the technical details of the `Stream` trait so this question might be way off.

**Comment by mcronce** [comment]
I'm hitting the same issue.  It seems like, to put a super fine point on it, the problem is that `DeserializeRecordsIntoStream` isn't `Send`.  I'm messing around with trying to come up with a solution in my own code, if I come up with anything that works I'll report back

**Comment by mcronce** [comment]
This patch for `csv-async` itself does get it working, but obviously restricts you to only using `Send + Sync` readers under the hood:
```diff
diff --git a/src/async_readers/ades_futures.rs b/src/async_readers/ades_futures.rs
index 440e82d..8be6dfe 100644
--- a/src/async_readers/ades_futures.rs
+++ b/src/async_readers/ades_futures.rs
@@ -48,7 +48,7 @@ impl AsyncReaderBuilder {
     ///     Ok(())
     /// }
     /// ```
-    pub fn create_deserializer<R: io::AsyncRead + std::marker::Unpin>(&self, rdr: R) -> AsyncDeserializer<R> {
+    pub fn create_deserializer<R: io::AsyncRead + std::marker::Unpin + Send + Sync>(&self, rdr: R) -> AsyncDeserializer<R> {
         AsyncDeserializer::new(self, rdr)
     }
 }
@@ -141,7 +141,7 @@ pub struct AsyncDeserializer<R>(AsyncReaderImpl<R>);
 
 impl<'r, R> AsyncDeserializer<R>
 where
-    R: io::AsyncRead + std::marker::Unpin + 'r,
+    R: io::AsyncRead + std::marker::Unpin + Send + Sync + 'r,
 {
     /// Create a new CSV reader given a builder and a source of underlying
     /// bytes.
diff --git a/src/async_readers/mod.rs b/src/async_readers/mod.rs
index 0480891..8c9c347 100644
--- a/src/async_readers/mod.rs
+++ b/src/async_readers/mod.rs
@@ -1695,7 +1695,7 @@ where
 /// type, and `D` refers to the type that this stream will deserialize a record into.
 pub struct DeserializeRecordsIntoStream<'r, R, D>
 where
-    R: io::AsyncRead + std::marker::Unpin
+    R: io::AsyncRead + std::marker::Unpin + Send + Sync
 {
     header_fut: Option<
         Pin<
@@ -1705,7 +1705,7 @@ where
                         Result<StringRecord>,
                         AsyncReaderImpl<R>,
                     )
-                > + 'r,
+                > + Send + Sync + 'r,
             >,
         >,
     >,
@@ -1719,7 +1719,7 @@ where
                         Option<StringRecord>,
                         StringRecord,
                     )
-                > + 'r,
+                > + Send + Sync + 'r,
             >,
         >,
     >,
@@ -1727,7 +1727,7 @@ where
 
 impl<'r, R, D: DeserializeOwned + 'r> DeserializeRecordsIntoStream<'r, R, D>
 where
-    R: io::AsyncRead + std::marker::Unpin + 'r
+    R: io::AsyncRead + std::marker::Unpin + Send + Sync + 'r
 {
     fn new(mut rdr: AsyncReaderImpl<R>) -> Self {
         let has_headers = rdr.has_headers();
@@ -1751,7 +1751,7 @@ where
 
 impl<'r, R, D: DeserializeOwned + 'r> Stream for DeserializeRecordsIntoStream<'r, R, D>
 where
-    R: io::AsyncRead + std::marker::Unpin + 'r
+    R: io::AsyncRead + std::marker::Unpin + Send + Sync + 'r
 {
     type Item = Result<D>;
 ```

**Comment by mcronce** [comment]
This patch gives you more flexibility by only marking the stream `Send + Sync` if the underlying reader is also `Send + Sync`, while allowing you to have non-`Send + Sync` streams...however, I have legit no idea how to go about proving whether or not it's sound, this is just mimicking what `Arc` does :)

```diff
diff --git a/src/async_readers/ades_futures.rs b/src/async_readers/ades_futures.rs
index 440e82d..335c3d2 100644
--- a/src/async_readers/ades_futures.rs
+++ b/src/async_readers/ades_futures.rs
@@ -139,6 +139,9 @@ impl AsyncReaderBuilder {
 #[derive(Debug)]
 pub struct AsyncDeserializer<R>(AsyncReaderImpl<R>);
 
+unsafe impl<R: Send + Sync> Send for AsyncDeserializer<R> {}
+unsafe impl<R: Send + Sync> Sync for AsyncDeserializer<R> {}
+
 impl<'r, R> AsyncDeserializer<R>
 where
     R: io::AsyncRead + std::marker::Unpin + 'r,
diff --git a/src/async_readers/mod.rs b/src/async_readers/mod.rs
index 0480891..8b98f47 100644
--- a/src/async_readers/mod.rs
+++ b/src/async_readers/mod.rs
@@ -1725,6 +1725,9 @@ where
     >,
 }
 
+unsafe impl<'r, R: io::AsyncRead + Unpin + Send + Sync, D> Send for DeserializeRecordsIntoStream<'r, R, D> {}
+unsafe impl<'r, R: io::AsyncRead + Unpin + Send + Sync, D> Sync for DeserializeRecordsIntoStream<'r, R, D> {}
+
 impl<'r, R, D: DeserializeOwned + 'r> DeserializeRecordsIntoStream<'r, R, D>
 where
     R: io::AsyncRead + std::marker::Unpin + 'r
```

**Comment by gwierzchowski** [comment]
@mcronce : Many thanks for comments - it looks reasonably.
I will try to look at it over weekend (this, or the latest next one) and add to the code maybe with some test.

**Comment by mcronce** [comment]
@gwierzchowski (and @joonnna) I have a branch with these changes (the 2nd set) here - https://github.com/mcronce/csv-async/tree/send-sync-deserializer - that I'm using in a couple places.  I'll make a PR if you want to mess with them.  You can use them as-is with `csv-async = {git = "https://github.com/mcronce/csv-async", branch = "send-sync-deserializer"}` in your `Cargo.toml` in case you need it for prototyping or anything.

Like I said, though, I have no idea how to prove soundness here :)

**Comment by gwierzchowski** [comment]
After some reflection I decided for more verbose approach shown in first post by @mcronce. Its more save and straightforward. I have checked that io Readers from most popular async frameworks: `tokio`, `async-std` and `smol` all implement Send+Sync, so it should not be a problem with such constraint. 
@mcronce, @joonnna - please test if code in branch `feature/SendDeserialize` works in your solution if you got chance.

I plan to merge in a week and publish in about 2-3 weeks after few more changes.
Many thanks for contribution!

**Comment by gwierzchowski** [comment]
Merged Send+Sync in #5

**Comment by sundy-li** [comment]
I think we should remove `Sync` trait, because without this `csv-async` can also work.

If this have less trait bounds, the more flexible users can use.
